### PR TITLE
Non-success status codes should not always result in a thrown exception

### PR DESCRIPTION
**Pull request by hlship**
There are cases where a valid request to GitHub may result in a 404, and dealing with that as a thrown exception, rather than a status code, is problematic.

Also, bumped many deps to latest, except where that causes a conflict with JDK 1.8.

### CODE REVIEW

File: project.clj [hunk 1]
```diff
@@ -9,29 +9,29 @@
...
-                 [http-kit "2.5.3"]
+  :dependencies [[org.clojure/clojure "1.11.3"]
+                 [cheshire "5.13.0"]
+                 [clj-commons/clj-yaml "1.0.27"]
+                 [http-kit "2.8.0"]
                  [nubank/clj-github-app "0.2.1"]
-                 [nubank/state-flow "5.14.0"]
-                 [clj-commons/fs "1.6.310"]
+                 [nubank/state-flow "5.17.0"]
+                 [clj-commons/fs "1.6.311"]
```
Comment (phmeier-nubank) [inline project.clj:22]: Are the version bumps essential to this PR? Otherwise separate them out at least to a separate commit.

File: src/clj_github/httpkit_client.clj [hunk 2]
```diff
@@ -29,13 +29,35 @@
...
+(defn request
+  "Sends a synchronous request to GitHub, largely as a wrapper around HTTP Kit.
+
+  In addition to normal HTTP Kit keys in the request map, two additional keys are added.
+
+  :path - used to create the :url key for HTTP kit; this is the path relative to https://api.github.com.
+
+  :throw? - if true (the default), then non-success status codes result in a thrown exception.
+    If set to false, then the response is returned, regardless and the caller can decide what to do
+    with failure statuses."
```
Comment (phmeier-nubank) [inline src/clj_github/httpkit_client.clj]: :100: for adding better docs.

Maybe specific what `non-success` status codes are, just to be sure. You can refer to the `success-codes` function.

Also add that in case of success the response body is conditionally parsed into json as returned as a clojure data structure.

File: test/clj_github/httpkit_client_test.clj [hunk 3]
```diff
@@ -31,11 +31,18 @@
...
         (is (thrown-with-msg? clojure.lang.ExceptionInfo #"(?i)Request to GitHub failed"
                               (sut/request client {})))))
-    ;; in case of lower level errors, e.g. DNS lookup failue there will not be a status code
+    (testing "it does not throw with a failure status if so enabled"
+      (with-fake-http [{:headers {"Authorization" "Bearer token"
+                                  "Content-Type"  "application/json"}}
+                       {:status 404}]
+                      (is (match?
+                            {:status 404}
+                            (sut/request client {:throw? false})))))
```
Comment (phmeier-nubank) [inline test/clj_github/httpkit_client_test.clj:44]: You should also verify that the headers and body are not mangled.

**Review by phmeier-nubank — APPROVED** [review]
LGTM. Some improvements suggested but they should not hold up the release train.  

image

File: project.clj (hunk 1 above)
Comment (hlship) [reply project.clj:22]: When the cost of getting a commit merged is too high, you tend to put a lot into one PR.  These changes are not essential, but I tend to update dependencies as hygiene, when possible.

File: src/clj_github/httpkit_client.clj (hunk 2 above)
Comment (hlship) [reply src/clj_github/httpkit_client.clj]: Adding docs is what I do :tm:.

**Comment by hlship** [comment]
Thanks for the review; I'll make some changes and merge.
